pkg/util/platformworkloadidentity/platformworkloadidentityrolesbyversion.go

```diff
@@ -0,0 +1,81 @@
+package platformworkloadidentity
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+// PlatformWorkloadIdentityRolesByVersion is the interface that validates and obtains the version from an PlatformWorkloadIdentityRoleSetDocument.
+type PlatformWorkloadIdentityRolesByVersion interface {
+	GetPlatformWorkloadIdentityRolesByRoleName() map[string]api.PlatformWorkloadIdentityRole
+}
+
+// platformWorkloadIdentityRolesByVersionService is the default implementation of the PlatformWorkloadIdentityRolesByVersion interface.
+type platformWorkloadIdentityRolesByVersionService struct {
+	platformWorkloadIdentityRoles []api.PlatformWorkloadIdentityRole
+}
+
+func NewPlatformWorkloadIdentityRolesByVersion(ctx context.Context, oc *api.OpenShiftCluster, dbPlatformWorkloadIdentityRoleSets database.PlatformWorkloadIdentityRoleSets) (PlatformWorkloadIdentityRolesByVersion, error) {
+	if !oc.UsesWorkloadIdentity() {
+		return nil, nil
+	}
+
+	currentInstallVersion := oc.Properties.ClusterProfile.Version
+	currentOpenShiftVersion, err := version.ParseVersion(currentInstallVersion)
+	if err != nil {
+		return nil, err
+	}
+	currentMinorVersion := currentOpenShiftVersion.MinorVersion()
+	requiredMinorVersions := map[string]bool{currentMinorVersion: false}
+	platformWorkloadIdentityRoles := []api.PlatformWorkloadIdentityRole{}
+
+	docs, err := dbPlatformWorkloadIdentityRoleSets.ListAll(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo != nil {
```

Comment:
IMO, this whole file could benefit from some code comments for reviewers and readability in general. I can mostly follow the logic but for others who might not be aware of the assumptions we are making around upgradeableTo and what specifically we are validating I think it would help!

Reply:
Added a comment to the top of the function.